Userland/Utilities/json.cpp

```diff
@@ -121,33 +149,181 @@ void print(JsonValue const& value, int spaces_per_indent, int indent, bool use_c
+            start = index.value();
+        }
+        if (components.size() > 1 && !components.at(1).is_empty()) {
+            auto index = components.at(1).to_int();
+            if (!index.has_value())
+                return Error::from_string_view("Invalid end index"sv);
+            end = index.value();
+        }
+        if (components.size() > 2 && !components.at(2).is_empty()) {
+            auto optional_step = components.at(2).to_int();
+            if (!optional_step.has_value())
+                return Error::from_string_view("Invalid step"sv);
+            step = optional_step.value();
+
+            // Start and end have different default values if the step is negative
+            if (step < 0 && start == 0)
+                start = INT_MAX;
+            if (step < 0 && end == INT_MAX)
+                end = INT_MIN;
+        }
+        return make_ref_counted<SliceKey>(start, end, step);
+    }
+
+    Vector<JsonValue> select(JsonValue const& value) override
+    {
+        Vector<JsonValue> matches;
+        auto const& array = value.as_array();
+        if (m_step > 0) {
+            int start = max(0, normalize_index(m_start, array.size()));
+            int end = min(array.size(), normalize_index(m_end, array.size()));
+            for (int i = start; i < end; i += m_step)
+                matches.append(array.at(i));
+        } else {
+            int start = min(array.size() - 1, normalize_index(m_start, array.size()));
+            int end = max(-1, normalize_index(m_end, array.size()));
+            for (int i = start; i > end; i += m_step)
+                matches.append(array.at(i));
+        }
+        return matches;
+    }
+};
+
+ErrorOr<JsonValue> query(JsonValue const& value, Vector<StringView>& key_parts, size_t key_index)
+{
+    if (key_index == key_parts.size())
+        return value;
+    auto key_str = key_parts[key_index++];
+
+    if (is_single_value_key(key_str)) {
+        auto key = SingleValueKey::parse(key_str);
+        auto result = TRY(key->select(value));
+        return query(result, key_parts, key_index);
+    } else {
+        auto key = TRY(MultiValueKey::parse(key_str));
+        Vector<JsonValue> results;
+        for (auto const& match : key->select(value)) {
+            auto result = TRY(query(match, key_parts, key_index));
+            results.append(result);
         }
-        return JsonValue(JsonArray(matches));
+        return JsonValue(JsonArray(results));
     }
+}
+
+// Normalize negative indices, e.g. an index of -1 would wrap to the last element.
+int normalize_index(int i, size_t length) { return (i < 0) ? length + i : i; }
+
+bool is_single_value_key(StringView key)
+{
+    if (key == "*"sv || key == "[*]"sv)
+        return false;
+    if (key.starts_with('[') && key.ends_with(']')) {
+        auto contains_quote = key.contains('"') || key.contains('\'');
+        auto is_filter = key[1] == '?'; // filters aren't yet implemented
```

Comment:
If something is unimplemented, it should be marked as `TODO`

Reply:
I went ahead and removed the comment. I'm not sure it's really necessary.